drivers/spi/spi_nrf5.c

```diff
@@ -0,0 +1,339 @@
+	dev_data->busy = false;
+}
+
+static int transceive(const struct spi_config *spi_cfg,
+		      const struct spi_buf_set *tx_bufs,
+		      const struct spi_buf_set *rx_bufs,
+		      bool asynchronous,
+		      struct k_poll_signal *signal)
+{
+	struct spi_nrf5_data *dev_data = get_dev_data(spi_cfg->dev);
+	int error;
+
+	spi_context_lock(&dev_data->ctx, asynchronous, signal);
+
+	error = configure(spi_cfg);
+	if (error == 0) {
+		dev_data->busy = true;
+
+		spi_context_buffers_setup(&dev_data->ctx, tx_bufs, rx_bufs, 1);
+		spi_context_cs_control(&dev_data->ctx, true);
+
+		transfer_next_chunk(spi_cfg->dev);
+
+		error = spi_context_wait_for_completion(&dev_data->ctx);
+	}
+
+	spi_context_release(&dev_data->ctx, error);
+
+	return error;
+}
+
+static int spi_nrf5_transceive(const struct spi_config *spi_cfg,
+			       const struct spi_buf_set *tx_bufs,
+			       const struct spi_buf_set *rx_bufs)
+{
+	return transceive(spi_cfg, tx_bufs, rx_bufs, false, NULL);
+}
+
+#ifdef CONFIG_SPI_ASYNC
+static int spi_nrf5_transceive_async(const struct spi_config *spi_cfg,
+				     const struct spi_buf_set *tx_bufs,
+				     const struct spi_buf_set *rx_bufs,
+				     struct k_poll_signal *async)
+{
+	return transceive(spi_cfg, tx_bufs, rx_bufs, true, async);
+}
+#endif /* CONFIG_SPI_ASYNC */
+
+static int spi_nrf5_release(const struct spi_config *spi_cfg)
+{
+	struct spi_nrf5_data *dev_data = get_dev_data(spi_cfg->dev);
+
+	if (!spi_context_configured(&dev_data->ctx, spi_cfg)) {
+		return -EINVAL;
+	}
+
+	if (dev_data->busy) {
+		return -EBUSY;
+	}
+
+	spi_context_unlock_unconditionally(&dev_data->ctx);
+
+	return 0;
+}
+
+static const struct spi_driver_api spi_nrf5_api = {
+	.transceive = spi_nrf5_transceive,
+#ifdef CONFIG_SPI_ASYNC
+	.transceive_async = spi_nrf5_transceive_async,
+#endif
+	.release = spi_nrf5_release,
+};
+
+
+static void event_handler(const nrfx_spi_evt_t *p_event, void *p_context)
+{
+	if (p_event->type == NRFX_SPI_EVENT_DONE) {
+		struct device *dev = p_context;
+		struct spi_nrf5_data *dev_data = get_dev_data(dev);
+
+		spi_context_update_tx(&dev_data->ctx, 1, dev_data->chunk_len);
+		spi_context_update_rx(&dev_data->ctx, 1, dev_data->chunk_len);
+
+		transfer_next_chunk(dev);
+	}
+}
+
+static int spi_init(struct device *dev,
+		    u8_t sck_pin,
+		    u8_t mosi_pin,
+		    u8_t miso_pin,
+		    u8_t orc)
+{
+	const nrfx_spi_config_t config = {
+		.sck_pin   = sck_pin,
+		.mosi_pin  = mosi_pin,
+		.miso_pin  = miso_pin,
+		.ss_pin    = NRFX_SPI_PIN_NOT_USED,
+		.orc       = orc,
+		/* Default settings, configure() will set the proper ones. */
+		.frequency = NRF_SPI_FREQ_4M,
+		.mode      = NRF_SPI_MODE_0,
+		.bit_order = NRF_SPI_BIT_ORDER_MSB_FIRST,
+	};
+	nrfx_err_t result = nrfx_spi_init(&get_dev_config(dev)->spi,
+					  &config,
+					  event_handler,
+					  dev);
+	if (result != NRFX_SUCCESS) {
+		SYS_LOG_ERR("Failed to initialize device: %s",
+			    dev->config->name);
+		assert(false);
```

Comment:
Yes, it should return an error. But this value is currently [ignored](https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/device.c#L54).
Added a `return` statement, and a comment.

Reply:
Still: return a valid error code.

I have plans to check these and do something about it (I may introduce it in the relevant pr actually), so it's preferable that all drivers conforms to the signature, even if it's not handled right now. The day it will be: it will be seamless (no need to fix any drivers).